src/workerd/server/server.c++

```diff
@@ -1653,50 +1653,80 @@ class Server::WorkerService final: public Service,
-        actor == kj::none ? ExecutionModel::STATELESS : ExecutionModel::DURABLE_OBJECT;
-    auto workerTracer =
-        tracer->makeWorkerTracer(PipelineLogLevel::FULL, executionModel, kj::none /* scriptId */,
-            kj::none /* stableId */, kj::none /* scriptName */, kj::none /* scriptVersion */,
-            kj::none /* dispatchNamespace */, nullptr /* scriptTags */, kj::none /* entrypoint */);
-
     auto& channels = KJ_ASSERT_NONNULL(ioChannels.tryGet<LinkedIoChannels>());
 
-    auto tailWorkers = KJ_MAP(service, channels.tails) -> kj::Own<WorkerInterface> {
-      KJ_ASSERT(service != this, "A worker currently cannot log to itself");
-      // Caution here... if the tail worker ends up have a cirular dependency
-      // on the worker we'll end up with an infinite loop trying to initialize.
-      // We can test this directly but it's more difficult to test indirect
-      // loops (dependency of dependency, etc). Here we're just going to keep
-      // it simple and just check the direct dependency.
-      return service->startRequest({});
-    };
-
-    // When the tracer is complete, deliver the traces to both the parent
-    // and the tail workers. We do NOT want to attach the tracer to the
-    // tracer->onComplete() promise here because it is the destructor of
-    // the PipelineTracer that resolves the onComplete promise. If we attach
-    // the tracer to the promise the tracer won't be destroyed while the
-    // promise is still pending! Fortunately, the WorkerTracer we created
-    // will hold a strong reference to the worker tracer for as long as it
-    // is needed. As long as the WorkerTracer is still alive, the PipelineTracer
-    // will be. See below, we end up creating two references to the WorkerTracer,
-    // one held by the observer and one that will be passed to the IoContext.
-    // The PipelineTracer will be destroyed once both of those are freed.
-    waitUntilTasks.add(tracer->onComplete().then(
-        kj::coCapture([tailWorkers = kj::mv(tailWorkers)](
-                          kj::Array<kj::Own<Trace>> traces) mutable -> kj::Promise<void> {
-      for (auto& worker: tailWorkers) {
-        auto event = kj::heap<workerd::api::TraceCustomEventImpl>(
-            workerd::api::TraceCustomEventImpl::TYPE, mapAddRef(traces));
-        co_await worker->customEvent(kj::mv(event)).ignoreResult();
+    kj::Array<kj::Own<WorkerInterface>> legacyTailWorkers = nullptr;
+    kj::Array<kj::Own<WorkerInterface>> streamingTailWorkers = nullptr;
+    // If streaming tail workers is enabled, then we will initialize two lists:
+    // one with services that only export the tail or trace handler (legacy tail
+    // workers) and one that exports the tailStream handler. We'll check tailStreams
+    // first.
+    if (util::Autogate::isEnabled(util::AutogateKey::STREAMING_TAIL_WORKERS)) {
+      kj::Vector<kj::Own<WorkerInterface>> legacyList;
+      kj::Vector<kj::Own<WorkerInterface>> streamingList;
```

Comment:
Depending on the size of `channels.tails` you can at least reserve the capacity for these vectors by `channels.tails.size() / 2`

Reply:
eh, I'm not going to worry about it here. This is just the workerd local server at this point so it really doesn't need to be highly performance optimized. I think the internal server that uses this would benefit from a few more optimizations but here it's not that critical.